feat: TAPAS execution workflow

yuukiiwa opened the pull request:
## Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have updated corresponding READMEs (if applicable)
- [x] My code follows the templates/style guidelines of the repository
- [x] In- and output formats comply with APAeval specifications
- [x] No parameters or file names are hardcoded
- [x] Results, logs or other output is not commited to the repository
- [x] I have tested my code

faricazjj commented on execution_workflows/TAPAS/nextflow.config:
```diff
@@ -0,0 +1,131 @@
+
+manifest {
+  name = 'APAeval'
+  author = 'your name'
```
to be updated

yuukiiwa replied:
I have updated this

faricazjj commented on execution_workflows/TAPAS/README.md:
```diff
@@ -0,0 +1,11 @@
+# APAeval nextflow pilot run
```
readme to be updated

yuukiiwa replied:
I have updated this

faricazjj reviewed (commented):
Thanks @yuukiiwa !! :D I looked through the code and it looks good to me except a couple places to update! Also is it missing the example samplesheet.csv?

yuukiiwa commented:
Hi @faricazjj, thank you for reviewing!! I have updated the `README.md` and `nextflow.config`, and added the module+script for generating `tapas_quant.bed` for the quantification challenge.
Thanks!

SamBryce-Smith commented on execution_workflows/TAPAS/bin/make_tapas_ref.py:
```diff
@@ -0,0 +1,30 @@
+ ln=ln.split("\t")
+ transcript="NM_"+ln[3]
+ gene="NM_"+transcript2gene[ln[3]]
+ organized=[gene,transcript,ln[0],ln[5],ln[1],ln[2],ln[6],ln[7],ln[9],ln[10],ln[11]]
```
I removed stripping the 'chr' prefix in 676bceceb9aea862f2bb3ebe768b5d2471c1f38a. With my test data I found this meant the reference and the coverage files no longer matched, so I got empty output. Fixing this got me proper output iwth the test data
![tapas-chr-strip-mismatch-2022-07-20-16-14-17](https://user-images.githubusercontent.com/49978382/180019016-3c369c48-0926-4893-9f06-e6478621c055.png)

SamBryce-Smith reviewed (changes requested):
Hey @yuukiiwa , thanks for the work and bump on the PR review! I've been able to get a test run to finish (with a few minor commits I pushed to the branch) with [test data described here](https://github.com/iRNA-COSI/APAeval/pull/391#issue-1311053632). 
![tapas-test-run-remove-strip-2022-07-20 16-06-08](https://user-images.githubusercontent.com/49978382/180017010-5260e8af-9de0-4333-8304-1b4dcdfd5993.png)



Summary of comments:
- TAPAS should also qualify for the identification challenge. Also differential, but I don't think we have to implement that now as we probably aren't going to complete that challenge.
- My config had multiple samples, but only 1 output file is created. Same happens under `results/samtools/read_coverage.txt`. I suspect outputs from the previous sample are overwritten. Could you set up the pipeline to generate outputs per-sample?
- I'm also not sure if TAPAS actually generates TPMs for each polyA site, so it may not qualify for the TPM quantification challenge. It may do for the relative quant though. I'll need to do some more reading
- See specific comments for more requests. Fixing the PAS coordinates is a major one for me!

SamBryce-Smith commented on execution_workflows/TAPAS/modules/make_quant_bed.nf:
```diff
@@ -0,0 +1,23 @@
+
+    script:
+    """
+    make_quant_bed.py $tapas_quant_txt tapas_quant.bed
```
I changed this to refer to make_tapas_quant_bed.py in https://github.com/iRNA-COSI/APAeval/commit/e3aa867b78f7af858163790de51a5b5796a5455f as the previous script couldn't be found in bin/

SamBryce-Smith commented on execution_workflows/TAPAS/bin/make_tapas_ref.py:
```diff
@@ -0,0 +1,31 @@
+ ln=ln.split("\t")
+ transcript="NM_"+ln[3]
+ gene="NM_"+transcript2gene[ln[3]]
+ organized=[gene,transcript,ln[0].strip("chr"),ln[5],ln[1],ln[2],ln[6],ln[7],ln[9],ln[10],ln[11]]
```
Implemented in 676bceceb9aea862f2bb3ebe768b5d2471c1f38a

yuukiiwa commented on execution_workflows/TAPAS/README.md:
```diff
@@ -0,0 +1,55 @@
+Parameters used to run the TAPAS are specified in the nextflow.config file.
+Parameters relevant to the workflow itself are:
+- `input` - samplesheet.csv
+Note: TAPAS also allows users to build their own model during the TAPAS run, but this pipeline does not support building one's own model (one would have to build his/her own model by running TAPAS outside of this pipeline). Please refer to [TAPAS's documentation](https://github.com/luskry/TAPAS#options)
```
I got it removed

yuukiiwa commented on execution_workflows/TAPAS/README.md:
````diff
@@ -0,0 +1,55 @@
+```
+
+## Output & post-processing
+TAPAS outputs a gtf file containing both the TAPAS identified entries and the entries from the inputted gtf files. Hence, the post-processing of the TAPAS results filters out the gtf entries from the reference gtf file and only selects entries that indicates TAPAS as their source. The TAPAS entries are formatted into a bed file with the following columns:
````
I think this is my bad. These are specific to `Aptardi` instead of `TAPAS`. I have updated the `README` accordingly.

yuukiiwa commented on execution_workflows/TAPAS/modules/samtools.nf:
```diff
@@ -0,0 +1,29 @@
+    sorted_bam="sorted_"+"$bam"
+    accepted_reads_bam="accepted_reads.bam"
+    """
+    samtools sort $bam -o $sorted_bam
+    samtools index -b $sorted_bam
+    samtools view -b $sorted_bam > $accepted_reads_bam
+    samtools depth $accepted_reads_bam > read_coverage.txt    
```
I have removed the `samtools sort`, `index`, and `view`

yuukiiwa commented on execution_workflows/TAPAS/README.md:
```diff
@@ -0,0 +1,55 @@
+Note: TAPAS also allows users to build their own model during the TAPAS run, but this pipeline does not support building one's own model (one would have to build his/her own model by running TAPAS outside of this pipeline). Please refer to [TAPAS's documentation](https://github.com/luskry/TAPAS#options)
+
+### Running the TAPAS execution workflow
+- Download the test data [here](to be added). The current dataset is in a genomic region where there are enough reads to test TAPAS's PAS quantification functionality.
```
Yup

yuukiiwa commented on execution_workflows/TAPAS/bin/make_tapas_ref.py:
```diff
@@ -0,0 +1,31 @@
+ return transcript2gene
+transcript2gene=readGTF(gtf_path)
+
+tapas_ref_path="tapas_ref.txt"
```
I have updated this accordingly

yuukiiwa commented on execution_workflows/TAPAS/bin/make_tapas_quant_bed.py:
```diff
@@ -0,0 +1,19 @@
+    chromStart=sites[i]
+    chromEnd=str(int(chromStart)+1)
+   elif strand == '-':
+    chromEnd=sites[i]
+    chromStart=str(int(chromEnd)+1)
```
I have updated this to 
```
   elif strand == '-':
    chromEnd=sites[i]
    chromStart=str(int(chromEnd)-1)
```

yuukiiwa commented on execution_workflows/TAPAS/modules/tapas_quant.nf:
```diff
@@ -0,0 +1,27 @@
+
+  script:
+  tapas_quant_out="tapas_quant_"+"$read_coverage"
+  """
+  /APA_sites_detection -ref $tapas_ref -cov $read_coverage -o $tapas_quant_out -l 76
+  """
```
read_length is now inputted through `samplesheet.csv`

yuukiiwa commented on execution_workflows/TAPAS/README.md:
```diff
@@ -0,0 +1,55 @@
+
+- sample: name of the sample for logs (e.g control_replicate1)
+- bam: BAM input file for the sample
+- gtf: reference_annotation.gtf file
```
I have switched the `gtf` column to `read_length`

yuukiiwa commented on execution_workflows/TAPAS/samplesheet.csv:
```diff
@@ -0,0 +1,2 @@
+sample,bam,gtf
+sample,[/path/to]/SRR6795721.bam,[/path/to]/gencode.v39.annotation.gtf
```
`gtf` is inputted from the command line

faricazjj commented on execution_workflows/TAPAS/README.md:
```diff
@@ -0,0 +1,55 @@
+and the [README documentation](https://github.com/arefeen/TAPAS#tapas-tool-for-alternative-polyadenylation-site-analysis) was used as a reference
+to create the nextflow pipeline of this module.
+
+This workflow qualifies for the APAeval **quantification challenge only**.
```
Ohh good catch Sam! I agree with putting differential aside if we're not including it in the project. We can add in the readme that it does qualify but we havent implemented it yet :D

faricazjj commented on execution_workflows/TAPAS/README.md:
````diff
@@ -0,0 +1,54 @@
+```
+- Run the pipeline with the samplesheet.csv with the input paths updated using either docker or singularity containers
+```
+nextflow main.nf --input samplesheet.csv -profile <docker/singularity>
````
I saw gtf as one of the two mandatory options in nextflow.config, does that mean we are also specifying the gtf file in the same command line here? If so we should update this command line example to include --gtf

yuukiiwa replied:
I have updated this with `--gtf`

faricazjj commented on execution_workflows/TAPAS/bin/check_samplesheet.py:
```diff
@@ -0,0 +1,107 @@
+        out_dir = os.path.dirname(file_out)
+        make_dir(out_dir)
+        with open(file_out, "w") as fout:
+            fout.write(",".join(['sample','bam','gtf']) + "\n")
```
gtf -> read_length?

yuukiiwa replied:
I have changed this

yuukiiwa replied:
I have changed this

faricazjj commented on execution_workflows/TAPAS/modules/check_samplesheet.nf:
```diff
@@ -0,0 +1,26 @@
+}
+
+def get_sample_info(LinkedHashMap sample) {
+    return [ sample.sample, sample.bam, sample.gtf ]
```
gtf -> read_length?

yuukiiwa replied:
I have changed this

faricazjj reviewed (commented):
Thank you for making the latest changes @yuukiiwa and @SamBryce-Smith!! looks good to me, I think there are just some minor changes for variable names, readme, and a couple parameters and output file names to be made @yuukiiwa :D

yuukiiwa commented on execution_workflows/TAPAS/README.md:
```diff
@@ -0,0 +1,54 @@
+and the [README documentation](https://github.com/arefeen/TAPAS#tapas-tool-for-alternative-polyadenylation-site-analysis) was used as a reference
+to create the nextflow pipeline of this module.
+
+This workflow qualifies for the APAeval **quantification challenge only**.
```
I have changed this

yuukiiwa commented on execution_workflows/TAPAS/README.md:
```diff
@@ -0,0 +1,54 @@
+### Running the TAPAS execution workflow
+- Download the test data [here](to be added). The current dataset is in a genomic region where there are enough reads to test TAPAS's PAS quantification functionality.
+
+- Update the samplesheet.csv with the full path to the downloaded bam, gtf, and fasta files.
```
I have changed this

yuukiiwa commented on execution_workflows/TAPAS/README.md:
````diff
@@ -0,0 +1,54 @@
+
+- Update the samplesheet.csv with the full path to the downloaded bam, gtf, and fasta files.
+```
+sample,bam,gtf
````
I have changed this

yuukiiwa commented on execution_workflows/TAPAS/bin/check_samplesheet.py:
```diff
@@ -0,0 +1,107 @@
+def check_samplesheet(file_in, file_out):
+    """
+    This function checks that the samplesheet follows the following structure:
+    sample,bam,gtf
```
I have changed this

yuukiiwa commented on execution_workflows/TAPAS/bin/check_samplesheet.py:
```diff
@@ -0,0 +1,107 @@
+                    print_error("gtf does not have extension '.gtf'", 'Line', line)
+
+            ## Create sample mapping dictionary = {group: {replicate : [ barcode, input_file, genome, gtf, is_transcripts ]}}
+            sample_info = [ sample, bam, gtf ]
```
I have changed this

yuukiiwa commented on execution_workflows/TAPAS/bin/check_samplesheet.py:
```diff
@@ -0,0 +1,107 @@
+
+        ## Check header
+        MIN_COLS = 3
+        HEADER = ['sample','bam','gtf']
```
I have changed this

yuukiiwa commented on execution_workflows/TAPAS/subworkflows/input_check.nf:
```diff
@@ -0,0 +1,22 @@
+        .set { ch_sample }
+
+    emit:
+    ch_sample // [ sample, barcode, fasta, gtf, is_transcripts, annotation_str ]
```
I have changed this

yuukiiwa commented on execution_workflows/TAPAS/modules/make_beds.nf:
```diff
@@ -0,0 +1,23 @@
+
+process MAKE_BEDS {
+    tag "$sample"
+    publishDir "${params.outdir}/tapas/tapas_quant/", mode: params.publish_dir_mode
```
I have updated `modules.config`

faricazjj reviewed (approved):
Thank you so much @yuukiiwa !!! LGTM :D I *think* @SamBryce-Smith has to approve this too before github lets us merge this because there were changes requested

faricazjj commented:
@yuukiiwa I was trying to look at the results and notice that the output bed files are not published hence not found in results folder. I modified make_beds.nf to publish the outputs so now they're showing up in outdir

However one thing is, I notice that the bed files outputted for identification and quantification are both empty and also checked that the other files outputted are non empty except for tapas_quant_read_coverage.txt that's also empty, and I assume this is why the identification and quantification output bed files are empty since their results are extracted from this txt file. I spent some time looking into this already but haven't found anything. Do you know what might be happening? Is this also the case with your test data if you're using a different one?

yuukiiwa commented:
Thanks so much for the changes, @faricazjj!! I have been using a full-size bam file for testing TAPAS to get lines from the `bed` file:
![Screenshot 2022-09-08 at 8 32 09 AM](https://user-images.githubusercontent.com/41866052/189007964-cc7901da-20bf-4b1c-94c2-aa011b0a9e92.png)

faricazjj commented on execution_workflows/TAPAS/bin/make_tapas_ref.py:
```diff
@@ -25,6 +25,6 @@ def readGTF(gtf_path):
  transcript="NM_"+ln[3]
  gene="NM_"+transcript2gene[ln[3]]
- organized=[gene,transcript,ln[0],ln[5],ln[1],ln[2],ln[6],ln[7],ln[9],ln[10],ln[11]]
+ organized=[gene,transcript,ln[0].strip('chr'),ln[5],ln[1],ln[2],ln[6],ln[7],ln[9],ln[10],ln[11]]
```
Had a discussion with @mrgazzara about this since I'm not too familiar with whether it's safe to strip 'chr' or if it's better to have an alternate set of data without 'chr' prefix. I faced a similar issue with Dapars back then because the tool needed 'chr' prefix but our data didn't have it, but we ended up having an alternate test data with 'chr' prefix instead of manually adding 'chr' in dapars worfklow :p I'll refer this to @mrgazzara and @SamBryce-Smith  to review this since I'm not sure! :D

SamBryce-Smith replied:
I agree with this, I wouldn't remove the 'chr' prefix at all. In past runs with test data I have got output in the BED files with TAPAS (which suggests we don't need to remove it to get TAPAS to run), however this time I was unsuccessful, probably because this meant the chr names in the annotation file no longer matched those in the BAM file

yuukiiwa replied:
I tried running `TAPAS` with the 'chr', which gave me empty outputs, which is why I'm stripping it off as this only outputs an intermediate file.

faricazjj replied:
does this mean we should just have an alternate set of data *without* chr prefix for TAPAS? and maybe add in the readme as well that TAPAS needs to run without 'chr' prefix in all the dataset?

yuukiiwa replied:
Now, the workflow requires the input gtf file to have its `chr` prefix stripped

faricazjj reviewed (approved):
Thanks so much @yuukiiwa !! Everything looks good!

SamBryce-Smith commented on execution_workflows/TAPAS/modules/samtools.nf:
```diff
@@ -0,0 +1,24 @@
+    output:
+    tuple val(sample), path("*.txt"), val(read_length), emit: read_coverage
+
+    script:
+    """
+    samtools depth $bam > read_coverage.txt    
+    """
+}
```
Looks like this doesn't create a sample-specific coverage file, which is what we'd want when running tapas sample-by-sample. By that I mean I think we need something like:
```
script:
coverage_file="$sample" + "_read_coverage.txt"
"""
samtools depth $bam > $coverage_file
"""
```

I only see one output file in the results directory, which suggests the read_coverage.txt is overwritten when the pipeline is run (and I have no idea to which sample it corresponds).

![image](https://user-images.githubusercontent.com/49978382/190098775-13d3feb7-6dc0-4a60-beed-87ab7794176d.png)

yuukiiwa replied:
I have changed this

SamBryce-Smith commented on execution_workflows/TAPAS/bin/make_tapas_beds.py:
```diff
@@ -0,0 +1,24 @@
+   if strand == '+':
+    chromStart=sites[i]
+    chromEnd=str(int(chromStart)+1)
+   elif strand == '-':
+    chromEnd=sites[i]
+    chromStart=str(int(chromEnd)-1)
```
I need to double check this but this might introduce a shift of one 'to the left' for coordinates on the minus strand. Start coordinates are included in the range, ends are excluded. If sites[i] is the actual coordinate reported but it's set as the 'End' coordinate, it won't be considered as the single nucleotide (but one to the left will be). This is assuming that TAPAS's predictions are reported in 'zero based' convention (which I think I'd expect if you are pass TAPAS a BED as input

faricazjj replied:
saw that the input to tapas quant is tapas ref where the coordinates are taken from a bed file, so i think it's right that tapas output would be following the input bed file which is zero based. Does that mean we would treat both strands the same? In which we would have
chromStart=sites[i]
chromEnd=str(int(chromStart)+1)

I always get confused of coordinates and strands :,)

SamBryce-Smith replied:
I've dug into this a little deeper with some IGV browsing. I've concluded that the minus strand coordinates are shifted 1 to the left.

Example on the minus strand where TAPAS calls the distal site as the end of the reference transcript. With current code, the identification files (bottom two) are shifted one to the left of the final nucleotide of the annotated exon:

![tapas-one-off-error-from-reference-igv_snapshot](https://user-images.githubusercontent.com/49978382/192761394-402bf705-b6f8-4617-9b31-cac730c0a530.png)

Example on the plus strand (chr11:69,653,277-69,653,316) where TAPAS calls the distal site as the end of the reference transcript. Here can see identification files get the final nucleotide/PAS spot on (as suspected above):
![tapas-plus-strand-correct](https://user-images.githubusercontent.com/49978382/192762520-71862f49-0f3a-4af5-aa98-524d90e3252d.png)

@faricazjj is absolutely correct - the solution to correct the minus strand PAS coordinates is to set the reported coordinate as chromStart and +1 to get chromEnd. This means the procedure is identical regardless of the strand

yuukiiwa replied:
I have updated the coordinates with what [Farica suggested](https://github.com/iRNA-COSI/APAeval/pull/330#discussion_r972147372), where the + and - strands are treated the same as follows:
```
chromStart=sites[i]
chromEnd=str(int(chromStart)+1)
```

SamBryce-Smith reviewed (commented):
Hey @yuukiiwa thanks for the changes just a few quick comments! Agree with Farica's points and have a few more points to clarify. See review comments below for more details :)

faricazjj reviewed (approved):
Thanks so much @yuukiiwa !! everything looks good except for the question about coordinates that i'm not sure about, will approve for now and wait for @SamBryce-Smith's input!

mrgazzara commented:
I think things are looking good. I am still working on exactly what's going wrong with the +/- 1 on the different strands and how to handle these properly to get a proper `bed` file for SWFs

SamBryce-Smith reviewed (commented):
Hi @yuukiiwa delayed reply again apologies :) I've made a few additional clarifying comments, particularly on the 'chr' prefix & minus strand coordinate reporting questions (see specific review comments for details).

1 general thing - Maybe I've lost track but was under impression we'd also get relative quant output from the workflow? When I ran it with my test data I only got identification outputs. I checked the `nextflow.config` and didn't see any switches to turn on/off outputting relative quantification. I also looked at `bin/make_tapas_beds.py` and there are definitely lines/code to output the relative quantification files.

yuukiiwa commented on execution_workflows/TAPAS/README.md:
```diff
@@ -0,0 +1,57 @@
+- `relative_quantification_bed_suffix`
+
+### Running the TAPAS execution workflow
+- Download the test data [here](to be added). The current dataset is in a genomic region where there are enough reads to test TAPAS's PAS quantification functionality.
```
Updated the README

yuukiiwa commented on execution_workflows/TAPAS/README.md:
````diff
@@ -0,0 +1,57 @@
+```
+nextflow main.nf --input samplesheet.csv --gtf [/path/to/gtf] -profile <docker/singularity>
+```
+- Note: please make sure that the provided GTF file's chromosome names are without the `chr` prefix. Please strip the `chr` prefix off prior to running the workflow if the GTF file you use has a `chr` prefix in its chromosome names.
````
I have updated the note statement

yuukiiwa commented:
Thanks for the suggestions, @SamBryce-Smith! The `*_tapas_quantification.bed` outputted from the ewf is output file for relative quantification

faricazjj reviewed (dismissed):
LGTM thank you so much @yuukiiwa !! :D

SamBryce-Smith reviewed (approved):
Thanks @yuukiiwa for the speedy changes! When I ran with your updated changes, I had a successful test run, albeit with a warning message. I still could not see output for the relative quantification challenge. 
![tapas-run-with-warning-message-2022-09-29 10-24-33](https://user-images.githubusercontent.com/49978382/193002018-bfc40ca9-19e6-41d8-8824-836bb13f28ba.png)
![image](https://user-images.githubusercontent.com/49978382/193002194-3821ffc2-78e4-4cec-83a8-09938b53db35.png)

When checking `nextflow.config`, I realised the parameter name is actually `relative_quantification_bed_suffix` and not `quantification_bed_suffix`. This explains the lack of rel quant output and the warning message. The `quantification_bed_suffix` is undefined so set as empty. This means the quantification output file does not have a '.bed' suffix, so nextflow doesn't capture it as output (to my understanding).

In 03f3f7bae I made a quick change to refer to the right parameter. I got a successful test run and now I see relative quantification BED files reported!! I checked those BEDs and from a quick glance everything seems to be in order.
![tapas-my-rel-quant-fix-2022-09-29 10-46-38](https://user-images.githubusercontent.com/49978382/193003025-7110d23a-2b8b-4c0a-90e2-e78483c84719.png)
![image](https://user-images.githubusercontent.com/49978382/193003109-b5ccda4f-d0b5-47cd-a055-1ce3a5170021.png)

-ve strand distal site PAS coordinates are now defined correctly too so that issue is resolved :)
![tapas-negative-fixed](https://user-images.githubusercontent.com/49978382/193003359-7a320867-a2dd-450e-9737-901a89bebe01.png)


On my end this is now good to go (there is 1 non-blocking minor comment about the README). @faricazjj you will unfortunately have to re-review/re-approve :)

Thanks again @yuukiiwa for all your hard work on this Ewf :sunglasses:

faricazjj reviewed (approved):
Thank you @SamBryce-Smith for the latest changes!! :D LGTM!
